packages/grid/_modules_/grid/hooks/features/useColumnReorder.tsx

```diff
@@ -0,0 +1,158 @@
+import * as React from 'react';
+import { ColDef } from '../../models/colDef';
+import { useLogger } from '../utils';
+import { ApiRef } from '../../models';
+import {
+  DRAGEND,
+  COL_REORDER_START,
+  COL_REORDER_DRAG_OVER,
+  COL_REORDER_DRAG_OVER_HEADER,
+  COL_REORDER_DRAG_ENTER,
+  COL_REORDER_STOP,
+} from '../../constants/eventsConstants';
+import {
+  HEADER_CELL_DROP_ZONE_CSS_CLASS,
+  HEADER_CELL_DRAGGING_CSS_CLASS,
+} from '../../constants/cssClassesConstants';
+
+export interface CursorCoordinates {
+  x: number;
+  y: number;
+}
+
+const CURSOR_MOVE_DIRECTION_LEFT = 'left';
+const CURSOR_MOVE_DIRECTION_RIGHT = 'right';
+
+const reorderColDefArray = (
+  columns: ColDef[],
+  newColIndex: number,
+  oldColIndex: number,
+): ColDef[] => {
+  const columnsClone = columns.slice();
+
+  columnsClone.splice(newColIndex, 0, columnsClone.splice(oldColIndex, 1)[0]);
+
+  return columnsClone;
+};
+
+const getCursorMoveDirectionX = (currentCoordinates, nextCoordinates) => {
+  return currentCoordinates.x <= nextCoordinates.x
+    ? CURSOR_MOVE_DIRECTION_RIGHT
+    : CURSOR_MOVE_DIRECTION_LEFT;
+};
+
+const hasCursorPositionChanged = (
+  currentCoordinates: CursorCoordinates,
+  nextCoordinates: CursorCoordinates,
+): boolean =>
+  currentCoordinates.x !== nextCoordinates.x || currentCoordinates.y !== nextCoordinates.y;
+
+export const useColumnReorder = (columnsRef: React.RefObject<HTMLDivElement>, apiRef: ApiRef) => {
+  const logger = useLogger('useColumnReorder');
+
+  const dragCol = React.useRef<ColDef | null>(null);
+  const dragColNode = React.useRef<HTMLElement | null>(null);
+  const cursorPosition = React.useRef<CursorCoordinates>({
+    x: 0,
+    y: 0,
+  });
+  const removeDnDStylesTimeout = React.useRef<number>();
+
+  const handleDragEnd = React.useCallback((): void => {
+    logger.debug(`End dragging col ${dragCol.current!.field}`);
+    apiRef.current.publishEvent(COL_REORDER_STOP);
+
+    clearTimeout(removeDnDStylesTimeout.current);
+
+    columnsRef.current!.classList.remove(HEADER_CELL_DROP_ZONE_CSS_CLASS);
+    dragColNode.current!.parentElement!.classList.remove('MuiDataGrid-colCellMoving');
+    dragColNode.current!.removeEventListener(DRAGEND, handleDragEnd);
+    dragCol.current = null;
+    dragColNode.current = null;
+  }, [columnsRef, apiRef, logger]);
+
+  const handleDragStart = React.useCallback(
+    (col: ColDef, currentTarget: HTMLElement): void => {
+      logger.debug(`Start dragging col ${col.field}`);
+      apiRef.current.publishEvent(COL_REORDER_START);
+
+      dragCol.current = col;
+      dragColNode.current = currentTarget;
+      dragColNode.current.addEventListener(DRAGEND, handleDragEnd, { once: true });
+      dragColNode.current.classList.add(HEADER_CELL_DRAGGING_CSS_CLASS);
+      dragColNode.current.parentElement!.classList.add('MuiDataGrid-colCellMoving');
+      removeDnDStylesTimeout.current = window.setTimeout(() => {
```

Comment:
Should we use `window.setTimeout` or just `setTimeout`

Reply:
window.setTimeout disambiguate between the node.js and browser API.

Reply:
Yes, it was `setTimeout` before but @oliviertassinari change it to `window.setTimeout` because `setTimeout` requires a typing of `NodeJS.Timeout`.

Reply:
I suspect that TypeScript is confused because of the import of the events module of node.js. With webpack v5, it will no longer work. They have stopped to polyfill Node.js modules. So in theory, we will get reports that the grid is broken as more people upgrade to webpack v5 (release a few days ago).

Reply:
removing the `, 0` at the end seems to fix the problem so the `window.` prefix is no longer needed.